pylearn2/models/tests/test_convelemwise_cost.py

```diff
@@ -0,0 +1,172 @@
+                                num_channels=1),
+        layers=[ConvElemwise(layer_name='conv',
+                             nonlinearity=conv_nonlinearity,
+                             output_channels=output_channels,
+                             kernel_shape=shape,
+                             pool_shape=[1, 1],
+                             pool_stride=shape,
+                             irange=1.0)],
+        batch_size=batch_size
+    )
+
+    X = conv_model.get_input_space().make_theano_batch()
+    Y = conv_model.get_target_space().make_theano_batch()
+    Y_hat = conv_model.fprop(X)
+    g = theano.function([X], Y_hat)
+
+    # Construct an equivalent MLP which gives the same output
+    # after flattening both.
+    mlp_model = MLP(
+        layers=[mlp_nonlinearity(dim=output_channels,
+                                 layer_name='mlp',
+                                 irange=1.0)],
+        batch_size=batch_size,
+        nvis=nvis
+    )
+
+    W, b = conv_model.get_param_values()
+
+    W = W.astype('float32')
+    b = b.astype('float32')
+
+    W_mlp = np.zeros(shape=(output_channels, nvis))
+    for k in range(output_channels):
+        W_mlp[k] = W[k, 0].flatten()[::-1]
+    W_mlp = W_mlp.T
+    b_mlp = b.flatten()
+
+    W_mlp = W_mlp.astype('float32')
+    b_mlp = b_mlp.astype('float32')
+
+    mlp_model.set_param_values([W_mlp, b_mlp])
+
+    X1 = mlp_model.get_input_space().make_theano_batch()
+    Y1 = mlp_model.get_target_space().make_theano_batch()
+    Y1_hat = mlp_model.fprop(X1)
+    f = theano.function([X1], Y1_hat)
+
+    # Check that the two models give the same output
+    assert np.linalg.norm(f(x_mlp).flatten() - g(x).flatten()) < 10**-3
+
+    # Check that the two models have the same costs:
+    mlp_cost = theano.function([X1, Y1], mlp_model.cost(Y1, Y1_hat))
+    conv_cost = theano.function([X, Y], conv_model.cost(Y, Y_hat))
+
+    assert np.linalg.norm(conv_cost(x, y) - mlp_cost(x_mlp, y_mlp)) < 10**-3
+
+
+def check_unimplemented_case(conv_nonlinearity):
+    """Check a ConvNonlinearity does not have a `cost` method.
+
+    If the `cost` method gets implemented in the future, it should
+    be checked against a reference dense implementation.
+
+    Parameters
+    ----------
+    conv_nonlinearity: subclass of `ConvNonlinearity`
+        The non-linearity to provide to a `ConvElemwise` layer.
+    """
+
+    conv_model = MLP(
+        input_space=Conv2DSpace(shape=[1, 1],
+                                axes=['b', 0, 1, 'c'],
+                                num_channels=1),
+        layers=[ConvElemwise(layer_name='conv',
+                             nonlinearity=conv_nonlinearity,
+                             output_channels=1,
+                             kernel_shape=[1, 1],
+                             pool_shape=[1, 1],
+                             pool_stride=[1, 1],
+                             irange=1.0)],
+        batch_size=1
+    )
```

Comment:
This currently fails, we should take code from the other `check_` method instead.
Also, we should check that the output is the same, even if the cost is not implemented.